include/triton/core/tritonbackend.h

```diff
@@ -1359,6 +1360,86 @@ TRITONBACKEND_ISPEC TRITONSERVER_Error* TRITONBACKEND_ModelInstanceExecute(
     const uint32_t request_count);
 
 
+/// [BETA] Below is an experimental feature and the interface is subject to
+/// changes.
+///
+
+typedef enum TRITONBACKEND_attributetype_enum {
+  TRITONBACKEND_ATTRIBUTE_PREFERRED_INSTANCE_GROUP,
+} TRITONBACKEND_AttributeType;
+
+/// Query the backend for different model attributes. This function is optional,
+/// a backend is not required to implement it. This function is called when
+/// Triton requires further backend / model information to perform operations.
+/// This function may be called multiple times within the lifetime of the
+/// backend (between TRITONBACKEND_Initialize and TRITONBACKEND_Finalize), and
+/// optional 'model' may be provided if it is called for a particular model.
+/// The backend may return error to indicate that the specific attribute does
+/// not apply, otherwise, 'backend_attributes' should be created and set to
+/// reflect such attribute and 'nullptr' should be returned. Triton will take
+/// the ownership of 'backend_attributes' if 'nullptr' is returned.
+///
+/// \param backend The backend.
+/// \param model The model. 'nullptr' may be provided if no model is associated.
+/// \param type The attribute to query about.
+/// \param in_params The array of input parameters as a complement of the
+/// attribute inquiry.
+/// \param in_count The number of input parameters.
+/// \param backend_attributes The backend attribute. See
+/// TRITONBACKEND_BackendAttribute on how it should be set for different 'type'.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONBACKEND_ISPEC TRITONSERVER_Error* TRITONBACKEND_GetBackendAttribute(
+    TRITONBACKEND_Backend* backend, TRITONBACKEND_Model* model,
```

Comment:
Perhaps we can break this down further to `TRITONBACKEND_GetBackendAttribute` and  `TRITONBACKEND_GetModelAttribute`? I think we should also be more clear on when we are going to call this callback function. For the two use cases that we currently have in mind, Triton will call `TRITONBACKEND_GetBackendAttribute` once after the `TRIITONBACKEND_Initialize` and another time after `TRITONBACKEND_ModelInitialize` for each model.

Reply:
@GuanLuo can you reply to this comment? I think using the same "TRITONBACKEND_BackendAttribute" struct for both model attributes and backend attributes can be confusing.

Reply:
I dropped the reference to the model at this point. It could be useful to inquiry attributes at model level, but I haven't sorted out the timing of calling such function. For this particular case (preferred instance group), I think we need to inquiry and set the field in model config before auto-complete (so before model init)